### PR TITLE
[Completeness Sync] added completeness sync functionality as part of metadata sync and data sync.

### PR DESCRIPTION
**Pull request by srikanthkatta2207**
As part of this we have added functionality to sync data element's completeness status as part of data values sync. It syncs completeness as part of both data sync job and metadata sync job.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/DefaultCompleteDataSetRegistrationExchangeService.java [hunk 1]
```diff
@@ -217,6 +222,11 @@ public void writeCompleteDataSetRegistrationsJson( ExportParams params, OutputSt
         cdsrStore.writeCompleteDataSetRegistrationsJson( params, out );
     }
 
+    @Override
+    public void writeCompleteDataSetRegistrationsJson(Date lastUpdated, OutputStream outputStream, IdSchemes idSchemes) {
+        cdsrStore.writeCompleteDataSetRegistrationsJson( lastUpdated, outputStream, idSchemes );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/DefaultCompleteDataSetRegistrationExchangeService.java]: Code style, bracket on new line

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/DefaultCompleteDataSetRegistrationExchangeService.java [hunk 2]
```diff
@@ -507,6 +519,21 @@ private int batchImport( CompleteDataSetRegistrations completeRegistrations, Imp
                 validateStoredBy( storedBy, i18n );
                 storedBy = StringUtils.isBlank( storedBy ) ? currentUser : storedBy;
 
+                lastUpdatedBy = cdsr.getLastUpdatedBy();
+
+                if (lastUpdatedBy == null) {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/DefaultCompleteDataSetRegistrationExchangeService.java]: Code style, brackets on new line

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java [hunk 3]
```diff
@@ -121,10 +129,64 @@ public void writeCompleteDataSetRegistrationsJson( ExportParams params, OutputSt
         IOUtils.closeQuietly( outputStream );
     }
 
+    @Override
+    public void writeCompleteDataSetRegistrationsJson(Date lastUpdated, OutputStream outputStream, IdSchemes idSchemes) {
+        String dsScheme = idSchemes.getDataSetIdScheme().getIdentifiableString().toLowerCase();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java]: Code style, space between ( ) and arguments.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java [hunk 4]
```diff
@@ -116,10 +124,66 @@ public void writeCompleteDataSetRegistrationsJson( ExportParams params, OutputSt
...
+
+        jdbcTemplate.query( sql, new RowCallbackHandler()
+        {
+            @Override
+            public void processRow( ResultSet rs ) throws SQLException
+            {
+                CompleteDataSetRegistration completeDataSetRegistration = completeDataSetRegistrations.getCompleteDataSetRegistrationInstance();
+                PeriodType pt = PeriodType.getPeriodTypeByName( rs.getString( "ptname" ) );
+
+                User user = userService.getUser(rs.getInt( "lastupdatedby" ));
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java]:
```suggestion
                User user = userService.getUser( rs.getInt( "lastupdatedby" ) );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java [hunk 5]
```diff
@@ -116,10 +124,66 @@ public void writeCompleteDataSetRegistrationsJson( ExportParams params, OutputSt
...
+
+                completeDataSetRegistration.open();
+                completeDataSetRegistration.setDataSet( rs.getString( "dsid" ) );
+                completeDataSetRegistration.setPeriod( pt.createPeriod( rs.getDate( "pestart" ), calendar ).getIsoDate() );
+                completeDataSetRegistration.setOrganisationUnit( rs.getString( "ouid" ) );
+                completeDataSetRegistration.setAttributeOptionCombo( rs.getString( "aocid" ) );
+                completeDataSetRegistration.setDate( removeTime( rs.getString( "date" ) ) );
+                completeDataSetRegistration.setStoredBy( rs.getString( "storedby" ) );
+                completeDataSetRegistration.setLastUpdatedBy( user );
+                completeDataSetRegistration.setLastUpdated( removeTime(rs.getString( "lastupdated" )) );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java]:
```suggestion
                completeDataSetRegistration.setLastUpdated( removeTime( rs.getString( "lastupdated" ) ) );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java [hunk 6]
```diff
@@ -125,10 +126,71 @@ public void setStoredBy( String storedBy )
...
+    @Override
+    public void setCompleted( Boolean completed )
+    {
+        writeField( FIELD_IS_COMPLETED, completed );
+    }
+
+    @Override
+    public void setLastUpdatedBy( User lastUpdatedBy )
+    {
+        writeObject( lastUpdatedBy);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java]:
```suggestion
        writeObject( lastUpdatedBy );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java [hunk 7]
```diff
@@ -125,10 +126,71 @@ public void setStoredBy( String storedBy )
...
+            return;
+        }
+
+        try
+        {
+            generator.writeObjectField( fieldName, value );
+        }
+        catch ( IOException e )
+        {
+            // Intentionally ignored
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java]: Are you sure this Exception should not at the very least be logged?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java [hunk 8]
```diff
@@ -125,10 +126,71 @@ public void setStoredBy( String storedBy )
...
+    private void writeObject( User user )
+    {
+        if ( user == null )
+        {
+            return;
+        }
+
+        try
+        {
+            generator.writeFieldName( FIELD_LAST_UPDATED_BY);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java]:
```suggestion
            generator.writeFieldName( FIELD_LAST_UPDATED_BY );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java [hunk 9]
```diff
@@ -125,10 +126,71 @@ public void setStoredBy( String storedBy )
...
+
+            generator.writeStartObject();
+
+            generator.writeObjectField( "id", user.getUid());
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java]:
```suggestion
            generator.writeObjectField( "id", user.getUid() );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java [hunk 10]
```diff
@@ -125,10 +126,71 @@ public void setStoredBy( String storedBy )
...
+            generator.writeObjectField( "name", user.getDisplayName() );
+
+            generator.writeEndObject();
+
+        }
+        catch ( IOException e )
+        {
+            // Intentionally ignored
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/streaming/StreamingJsonCompleteDataSetRegistration.java]: Are you sure this should not even be logged?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DefaultDataValueSetService.java [hunk 11]
```diff
@@ -1360,7 +1360,7 @@ private void handleComplete( DataSet dataSet, Date completeDate, Period period,
         else
         {
             CompleteDataSetRegistration registration = new CompleteDataSetRegistration( dataSet, period, orgUnit,
-                attributeOptionCombo, completeDate, username );
+                    attributeOptionCombo, completeDate, username,completeAlready.getLastUpdatedBy(),completeDate,completeAlready.getCompleted());
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DefaultDataValueSetService.java]: Codestyle, space between parenthesis and parameter list

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/sync/MetadataSyncPreProcessor.java [hunk 12]
```diff
@@ -259,4 +259,14 @@ private MetadataVersion getLatestVersion( List<MetadataVersion> metadataVersionL
 
         return null;
     }
+
+    public void handleDataSetCompletenessPush(MetadataRetryContext context) {
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/sync/MetadataSyncPreProcessor.java]:
```suggestion
    public void handleDataSetCompletenessPush( MetadataRetryContext context ) {
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java [hunk 13]
```diff
@@ -68,14 +69,19 @@
 
     private final RestTemplate restTemplate;
 
+    private SynchronizationManager synchronizationManager;
+
 
     @Autowired
-    public DataValueSynchronization( DataValueService dataValueService, DataValueSetService dataValueSetService, SystemSettingManager systemSettingManager, RestTemplate restTemplate )
+    public DataValueSynchronization(DataValueService dataValueService, DataValueSetService dataValueSetService,
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java]:
```suggestion
    public DataValueSynchronization( DataValueService dataValueService, DataValueSetService dataValueSetService,
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java [hunk 14]
```diff
@@ -68,14 +69,19 @@
...
+                                    SystemSettingManager systemSettingManager, RestTemplate restTemplate,
+                                    SynchronizationManager synchronizationManager)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java]:
```suggestion
                                    SynchronizationManager synchronizationManager )
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java [hunk 15]
```diff
@@ -146,6 +152,34 @@ public SynchronizationResult syncDataValuesData()
...
+        final Clock clock = new Clock( log ).startClock().logTime( "Starting Completeness Synchronisation job." );
+
+        // ---------------------------------------------------------------------
+        // Set time for last success to start of process to make data saved
+        // subsequently part of next synch process without being ignored
+        // ---------------------------------------------------------------------
+        ImportSummary importSummary;
+        try {
+            importSummary = synchronizationManager.executeDataSetCompletenessPush();
+            if (checkSummaryStatus(importSummary, SyncEndpoint.COMPLETE_DATA_SET_REGISTRATIONS)) {
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java]:
```suggestion
            if ( checkSummaryStatus( importSummary, SyncEndpoint.COMPLETE_DATA_SET_REGISTRATIONS ) ) {
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java [hunk 16]
```diff
@@ -146,6 +152,34 @@ public SynchronizationResult syncDataValuesData()
...
+                clock.logTime( "SUCCESS! Completeness Synchronisation job is done. It took" );
+                setLastDataValueSynchronizationSuccess( new Date( clock.getStartTime() ) );
+                return SynchronizationResult.newSuccessResultWithMessage( "Completeness Synchronisation done. It took " + clock.getTime() + " ms." );
+            }
+
+        } catch (Exception ex) {
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java]:
```suggestion
        } catch ( Exception ex ) {
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DataSynchronizationJob.java [hunk 17]
```diff
@@ -83,6 +83,19 @@ public void execute( JobConfiguration jobConfiguration )
...
+        {
+            synchronizationManager.executeDataSetCompletenessPush();
+        }
+        catch ( RuntimeException ex )
+        {
+            notifier.notify( jobConfiguration, "completeness synch failed: " + ex.getMessage() );
+        }
+        catch ( WebMessageParseException e )
+        {
+            log.error("Error while executing data completeness task. "+ e.getMessage(), e );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DataSynchronizationJob.java]:
```suggestion
            log.error( "Error while executing data completeness task. "+ e.getMessage(), e );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java [hunk 18]
```diff
@@ -119,6 +127,96 @@ public AvailabilityStatus isRemoteServerAvailable()
...
+    }
+
+    private ImportSummary executeDataSetCompletenessPush( SystemInstance instance )  throws WebMessageParseException
+    {
+
+        final Date startTime = new Date();
+
+        final Date lastSuccessTime = getLastDataSynchSuccessFallback();
+
+        final int lastUpdatedCount = completeDataSetRegistrationService.getCompleteDataSetCountLastUpdatedAfter(lastSuccessTime );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java]:
```suggestion
        final int lastUpdatedCount = completeDataSetRegistrationService.getCompleteDataSetCountLastUpdatedAfter( lastSuccessTime );
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java [hunk 19]
```diff
@@ -119,6 +127,96 @@ public AvailabilityStatus isRemoteServerAvailable()
...
+        log.info( "Values: " + lastUpdatedCount + " since last synch success: " + lastSuccessTime );
+
+        if ( lastUpdatedCount == 0 )
+        {
+            setLastDataSynchSuccess( startTime );
+            log.debug( "Skipping completeness synch, no new or updated data values" );
+            return null;
+        }
+
+        log.info( "Values: " + lastUpdatedCount + " since last synch success: " + lastSuccessTime );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java]: You log the same information 2 times; lines 168 and 159.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java [hunk 20]
```diff
@@ -119,6 +127,96 @@ public AvailabilityStatus isRemoteServerAvailable()
...
+        {
+            request.getHeaders().setContentType( MediaType.APPLICATION_JSON );
+            request.getHeaders().add( HEADER_AUTHORIZATION, CodecUtils.getBasicAuthString( instance.getUsername(), instance.getPassword() ) );
+
+            completeDataSetRegistrationExchangeService.writeCompleteDataSetRegistrationsJson( lastSuccessTime, request.getBody(), new IdSchemes());
+        };
+
+        ResponseExtractor<ImportSummary> responseExtractor = new ImportSummaryResponseExtractor();
+        ImportSummary summary = null;
+        try
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java]:
```suggestion

        try
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceTest.java [hunk 21]
```diff
@@ -208,6 +205,9 @@ public void setUpTest()
         deG.setValueType( ValueType.TRUE_ONLY );
         dsA = createDataSet( 'A', new MonthlyPeriodType() );
         dsA.setCategoryCombo( categoryComboDef );
+
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetServiceTest.java]: Codestyle, spaces in parenthesis, please check whole file

File: dhis-2/dhis-support/dhis-support-jdbc/src/main/java/org/hisp/dhis/jdbc/batchhandler/CompleteDataSetRegistrationBatchHandler.java [hunk 22]
```diff
@@ -136,17 +139,20 @@ public boolean isInclusiveUniqueColumns()
             registration.getSource().getId(),
             registration.getAttributeOptionCombo().getId(),
             getLongDateString( registration.getDate() ),
-            registration.getStoredBy() );
+            registration.getStoredBy(),
+            registration.getLastUpdatedBy().getId(),
+                getLongDateString(registration.getLastUpdated()),
```
Comment (stian-sandvold) [inline dhis-2/dhis-support/dhis-support-jdbc/src/main/java/org/hisp/dhis/jdbc/batchhandler/CompleteDataSetRegistrationBatchHandler.java]:
```suggestion
            getLongDateString( registration.getLastUpdated() ),
```

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 23]
```diff
@@ -2073,7 +2074,7 @@ function getPreviousEntryField( field )
 // Data completeness
 // -----------------------------------------------------------------------------
 
-function registerCompleteDataSet()
+function registerCompleteDataSet(completedStatus)
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]:
```suggestion
function registerCompleteDataSet( completedStatus )
```

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 24]
```diff
@@ -2124,7 +2127,7 @@ function registerCompleteDataSet()
                 if( data && data.status == 'SUCCESS' )
                 {
                     $( document ).trigger( dhis2.de.event.completed, [ dhis2.de.currentDataSetId, params ] );
-                    disableCompleteButton();                    
+                    disableCompleteButton(params.isCompleted);
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]:
```suggestion
                    disableCompleteButton( params.isCompleted );
```

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 25]
```diff
@@ -2140,7 +2143,7 @@ function registerCompleteDataSet()
 	        	else // Offline, keep local value
 	        	{
                     $( document ).trigger( dhis2.de.event.completed, [ dhis2.de.currentDataSetId, params ] );
-	        		disableCompleteButton();
+	        		disableCompleteButton(params.isCompleted);
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]:
```suggestion
	        		disableCompleteButton( params.isCompleted );
```

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 26]
```diff
@@ -2198,7 +2201,7 @@ function undoCompleteDataSet()
         {
             dhis2.de.storageManager.clearCompleteDataSet( params );
             $( document ).trigger( dhis2.de.event.completed, [ dhis2.de.currentDataSetId, params ] );
-            disableUndoButton();         
+            disableCompleteButton(params);
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]:
```suggestion
            disableCompleteButton( params );
```

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 27]
```diff
@@ -2224,10 +2227,15 @@ function disableUndoButton()
     $( '#undoButton' ).attr( 'disabled', 'disabled' );
 }
 
-function disableCompleteButton()
+function disableCompleteButton(status)
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]:
```suggestion
function disableCompleteButton( status )
```

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 28]
```diff
@@ -2224,10 +2227,15 @@ function disableUndoButton()
...
 {
-    $( '#completeButton' ).attr( 'disabled', 'disabled' );
-    $( '#undoButton' ).removeAttr( 'disabled' );
+    if(status == true) {
```
Comment (stian-sandvold) [inline dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]:
```suggestion
    if( status == true ) {
```

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/CompleteDataSetRegistration.java [hunk 29]
```diff
@@ -86,14 +94,18 @@ public CompleteDataSetRegistration( DataSet dataSet, Period period, Organisation
     }
 
     public CompleteDataSetRegistration( DataSet dataSet, Period period, OrganisationUnit source,
-        CategoryOptionCombo attributeOptionCombo, Date date, String storedBy )
+        CategoryOptionCombo attributeOptionCombo, Date date, String storedBy, User lastUpdatedBy, Date lastUpdated,
+                                        Boolean isCompleted )
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/CompleteDataSetRegistration.java]: Codestyle

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/CompleteDataSetRegistration.java [hunk 30]
```diff
@@ -86,14 +94,18 @@ public CompleteDataSetRegistration( DataSet dataSet, Period period, Organisation
...
     {
         this.dataSet = dataSet;
         this.period = period;
         this.source = source;
         this.attributeOptionCombo = attributeOptionCombo;
         this.date = date;
         this.storedBy = storedBy;
+        this.lastUpdatedBy = lastUpdatedBy;
+        this.lastUpdated = lastUpdated;
+        this.completed = isCompleted;
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/CompleteDataSetRegistration.java]: Can we also name the argument "completed" ?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/hibernate/HibernateCompleteDataSetRegistrationStore.java [hunk 31]
```diff
@@ -167,4 +170,27 @@ public void deleteCompleteDataSetRegistrations( OrganisationUnit unit )
...
+    public int getCompleteDataSetCountLastUpdatedBetween( Date date )
+    {
+
+        if ( date == null )
+        {
+            throw new IllegalArgumentException( "date must be specified" );
+        }
+
+        Criteria criteria = sessionFactory.getCurrentSession()
+                .createCriteria( CompleteDataSetRegistration.class )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataset/hibernate/HibernateCompleteDataSetRegistrationStore.java]: Codestyle, indentation

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java [hunk 32]
```diff
@@ -151,19 +215,19 @@ private static String createQuery( ExportParams params )
...
-            "INNER JOIN organisationunit ou ON (cdsr.sourceid=ou.organisationunitid) " +
-            "INNER JOIN categoryoptioncombo aoc ON (cdsr.attributeoptioncomboid = aoc.categoryoptioncomboid) ";
+                "SELECT ds.${dsScheme} AS dsid, pe.startdate AS pe_start, pt.name AS ptname, ou.${ouScheme} AS ouid, " +
+                "aoc.${aocScheme} AS aocid, cdsr.storedby AS storedby, cdsr.date AS created " +
+                "FROM completedatasetregistration cdsr " +
+                "INNER JOIN dataset ds ON (cdsr.datasetid=ds.datasetid) " +
+                "INNER JOIN period pe ON (cdsr.periodid=pe.periodid) " +
+                "INNER JOIN periodtype pt ON (pe.periodtypeid=pt.periodtypeid) " +
+                "INNER JOIN organisationunit ou ON (cdsr.sourceid=ou.organisationunitid) " +
+                "INNER JOIN categoryoptioncombo aoc ON (cdsr.attributeoptioncomboid = aoc.categoryoptioncomboid) ";
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/dataset/JdbcCompleteDataSetRegistrationExchangeStore.java]: Can we avoid these extra tabs?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java [hunk 33]
```diff
@@ -146,6 +152,34 @@ public SynchronizationResult syncDataValuesData()
...
+                    "server is unavailable." );
+        }
+        final Clock clock = new Clock( log ).startClock().logTime( "Starting Completeness Synchronisation job." );
+
+        // ---------------------------------------------------------------------
+        // Set time for last success to start of process to make data saved
+        // subsequently part of next synch process without being ignored
+        // ---------------------------------------------------------------------
+        ImportSummary importSummary;
+        try {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java]: Brackets on new line

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java [hunk 34]
```diff
@@ -146,6 +152,34 @@ public SynchronizationResult syncDataValuesData()
         return SynchronizationResult.newFailureResultWithMessage( "DataValueSynchronization failed." );
     }
 
+    public SynchronizationResult syncCompleteness() {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/DataValueSynchronization.java]: Brackets on new line

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/CompleteDataSetRegistration.java (hunk 29 above)
Comment (phoenixTW) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataset/CompleteDataSetRegistration.java]: Could you provide some more input related to which code style you are referring too?

File: dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js [hunk 35]
```diff
@@ -1741,6 +1741,7 @@ function getOfflineDataValueJson( params )
 	json.complete = complete;
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-web/dhis-web-dataentry/src/main/webapp/dhis-web-dataentry/javascript/form.js]: Okay that makes sense.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java [hunk 36]
```diff
@@ -118,6 +126,94 @@ public AvailabilityStatus isRemoteServerAvailable()
         return SyncUtils.isRemoteServerAvailable( systemSettingManager, restTemplate );
     }
 
+    @Override
+    public ImportSummary executeDataSetCompletenessPush() throws WebMessageParseException {
+        AvailabilityStatus availability = isRemoteServerAvailable();
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/synch/DefaultSynchronizationManager.java]: Okay.

**Pull request merged**